### ISSUE TITLE
Megathread for Missing Features

### ISSUE THREAD

**Issue by adamjs** [ultralight-ux/Ultralight#178]
This issue will keep track of various things missing in Ultralight that are present in Chrome/Firefox.

* CSS Filter Effects (eg blur, hue, grayscale, contrast, etc.)
* CSS Text Shadow (dependent on compositing blurs in shader)
* WebGL (need to spawn offscreen EGL context or pipe into GPUDriver)
* Canvas API: (basic stroked/filled paths are working, need to finish everything else)
* Radial Gradients (just needs shader implementation)
* JavaScript Intl API / Localization
* WebSockets over SSL (WebSockets over HTTP work)
* WebAssembly
* Input Method Editor API (IME, used for Japanese, Korean, etc. input)
* Native Input: Select Elements (native widget dropdowns)
* Native Input: Radio / Checkbox Elements (native widget drawing)
* Native Input: Local File Selection
* Opening links with target="_blank" (need to add API for user to create new View)
* Platform clipboard API (cut, copy, paste)
* Platform drag and drop (dragging to/from native window)
* Animated GIF support (certain GIFs aren't rendering)
* JavaScript Alerts
* CSS Cursor Images

**Comment by PierBover** [comment]
You should probably add `<video>` to that list.

**Comment by adamjs** [comment]
Thanks, added ✅

**Comment by adamjs** [comment]
Thanks for the tip! Updated.

**Comment by alandefreitas** [comment]
It is possible to create a system tray icon for the application? Something [like that](https://doc.qt.io/qt-5/qtwidgets-desktop-systray-example.html). It's often useful for the user to close the window but to keep a deamon running in the background.

**Comment by talentlessguy** [comment]
What about adding dGPU support (NVIDIA for example)?

issue: #220

**Comment by gajus** [comment]
Is this being kept up to date?

**Comment by stel-la** [comment]
Hello! You sent an email to STREAM's public service host. This account is not monitored meaning that your email will be lost in the wonders of our servers...

**Comment by adamjs** [comment]
It's a bit out of date since most work the last two months has been in the 1.2 development branch (rebased off latest WebKit, changeset has 1M+ LOC). Going to push to public GitHub soon and update this thread.

**Comment by tamir1371** [comment]
Hey , how can i get a text value from a text input element.

**Comment by kapulkin** [comment]
Good day.
Is the list of features actual? I'am interested in WebRTC, is it not implemented yet?

**Comment by Jonatino** [comment]
@adamjs can you add support for webp images to this list?

**Comment by Jonatino** [comment]
@adamjs also if we can add support for icons for the windows?

**Comment by farshed** [comment]
@adamjs any update on v1.2? when can we expect it to be out?

**Comment by alandefreitas** [comment]
Is [supporting mobile devices](https://github.com/ultralight-ux/Ultralight/issues/42) still in the plans?

**Comment by SupinePandora43** [comment]
#254

**Comment by SupinePandora43** [comment]
`window.onKeypress` not work in `Browser` sample
example [electron-app](https://supinepandora43.github.io/electron-app/)
source: [#L40](https://github.com/SupinePandora43/electron-app/blob/master/src/App.tsx#L40)

**Comment by SupinePandora43** [comment]
![image](https://user-images.githubusercontent.com/36124472/84929338-1dad8100-b0e9-11ea-9300-e4d10fbd78ea.png)

**Comment by loeffel-io** [comment]
Gamepad API (!!)

**Comment by gtnardy** [comment]
looking forward for CSS Filters and HTML5 video and audio support!

**Comment by ryanerwin** [comment]
Looking forward to supporting media queries for dark mode, so that embedded Ultralight windows can toggle the light/dark mode automatically matching the Desktop OS Light/Dark mode setting.

For example: https://davy.dev/darkmode

This works with Epiphany (webkit) but not yet with Ultralight v1.2.0-beta.2

**Comment by antoineok** [comment]
hello, what about "loading webp image file" i got some errors with them

**Comment by SupinePandora43** [comment]
~~Chrome/Firefox~~ Chromium

**Comment by wolfield** [comment]
Please add https://github.com/ultralight-ux/Ultralight/issues/185 to the list. Chrome/Firefox had subpixel rendering since the beginning.

**Comment by HarryZalessky** [comment]
Is this list kept updated? video and audio support had been added... And what about WebRTC?? When is it coming?? I NEED IT!

**Comment by SupinePandora43** [comment]
> Is this list kept updated? video and audio support had been added... And what about WebRTC?? When is it coming?? I NEED IT!

It's not exactly done yet. 1.3 stable isn't released

**Comment by dzcpy** [comment]
> Is this list kept updated? video and audio support had been added... And what about WebRTC?? When is it coming?? I NEED IT!

Have video and audio been added?

**Comment by SupinePandora43** [comment]
> > Is this list kept updated? video and audio support had been added... And what about WebRTC?? When is it coming?? I NEED IT!
> 
> Have video and audio been added?

In latest sdk yes

**Comment by SupinePandora43** [comment]
YouTube says that fullscreen mode isn't supported

**Comment by dzcpy** [comment]
Is this project still actively developed / maintained?

**Comment by SupinePandora43** [comment]
> Is this project still actively developed / maintained?

Yes. 1.3.0 is currently in beta and adds audio/video support + more other features

**Comment by dzcpy** [comment]
Is there any estimation about when these features will be implemented? It's been too long.

**Comment by barnabwhy** [comment]
Any plans to add a way of disabling web security? Some things like vue just don't like file:// and complain about cors

**Comment by p0358** [comment]
> #include <Ultralight/MouseEvent.h>
```js
  ///
  /// The various mouse button types.
  ///
  enum Button {
    kButton_None = 0,
    kButton_Left,
    kButton_Middle,
    kButton_Right,
  };
  ```
No support for mouse side buttons (forward/back), and they are supposed to work. See sample here (press forward on the "Result" live sample): https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/buttons

So I cannot even catch them in WndProc and send to the view... (with WM_XBUTTONDOWN: https://stackoverflow.com/questions/67683479/process-mouse-back-and-forward-buttons-in-winforms-form-when-its-client-area-is)

**Comment by Kukulkano** [comment]
You might want to add **IndexedDB** and **local storage** support to that list (maybe include setting a dedicated storage path).

**Comment by augustpemberton** [comment]
MouseEvents don't seem to contain the movementX or movementY properties: 

https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/movementX
https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/movementY

**Comment by Steven-L-42** [comment]
how is it still 5 years later a problem to set a cursor to none or set it to custom in UL?
